src/v0/destinations/topsort/transform.js

```diff
@@ -0,0 +1,132 @@
+const {
+  InstrumentationError,
+  ConfigurationError,
+  getHashFromArrayWithDuplicate,
+} = require('@rudderstack/integrations-lib');
+const { ConfigCategory, mappingConfig, ECOMM_EVENTS_WITH_PRODUCT_ARRAY } = require('./config');
+const { constructPayload, simpleProcessRouterDest } = require('../../util');
+const {
+  constructItemPayloads,
+  createEventData,
+  isProductArrayValid,
+  getMappedEventName,
+  addFinalPayload,
+} = require('./utils');
+
+// Function to process events with a product array
+const processProductArray = (
+  products,
+  basePayload,
+  placementPayload,
+  topsortEvent,
+  apiKey,
+  finalPayloads,
+) => {
+  const itemPayloads = constructItemPayloads(products, mappingConfig[ConfigCategory.ITEM.name]);
+  itemPayloads.forEach((itemPayload) => {
+    const eventData = createEventData(basePayload, placementPayload, itemPayload, topsortEvent);
+    addFinalPayload(eventData, apiKey, finalPayloads);
+  });
+};
+
+// Function to process events with a single product or no product data
+const processSingleProduct = (
+  basePayload,
+  placementPayload,
+  message,
+  topsortEvent,
+  apiKey,
+  finalPayloads,
+  messageId,
+) => {
+  const itemPayload = constructPayload(message, mappingConfig[ConfigCategory.ITEM.name]);
+  const eventData = createEventData(basePayload, placementPayload, itemPayload, topsortEvent);
+
+  // Ensure messageId is used instead of generating a UUID for single product events
+  eventData.data.id = messageId;
+
+  // Add final payload with appropriate ID and other headers
+  addFinalPayload(eventData, apiKey, finalPayloads);
+};
+
+const responseBuilder = (message, { Config }) => {
+  const { apiKey, topsortEvents } = Config;
+  const { event, properties } = message;
+
+  // Parse Topsort event mappings
+  const parsedTopsortEventMappings = getHashFromArrayWithDuplicate(topsortEvents);
+  const mappedEventName = getMappedEventName(parsedTopsortEventMappings, event);
+
+  if (!mappedEventName) {
+    throw new InstrumentationError("Event not mapped in 'topsortEvents'. Dropping the event.");
+  }
+
+  const topsortEvent = mappedEventName;
+
+  // Construct base and placement payloads
+  const basePayload = constructPayload(message, mappingConfig[ConfigCategory.TRACK.name]);
+  const placementPayload = constructPayload(message, mappingConfig[ConfigCategory.PLACEMENT.name]);
+
+  // Check if the event involves a product array (using ECOMM_EVENTS_WITH_PRODUCT_ARRAY)
+  const isProductArrayAvailable =
+    ECOMM_EVENTS_WITH_PRODUCT_ARRAY.includes(event) && isProductArrayValid(event, properties);
+
+  const finalPayloads = [];
+
+  // Handle events based on the presence of a product array
+  if (isProductArrayAvailable) {
+    processProductArray(
+      properties.products,
+      basePayload,
+      placementPayload,
+      topsortEvent,
+      apiKey,
+      finalPayloads,
+    );
```

Comment:
Whenever there are more than 3 arguments to a function package all the arguments into a single object and pass it to avoid argument missmatch issue.
ref: https://www.youtube.com/watch?v=DBJ5-skcWzk